utils/data.cpp

```diff
@@ -0,0 +1,246 @@
+}
+
+Shape* ShapeData::extractNonText(QTextStream& qin, QString& shapeType, QList<int>& dimensionsList)
+{
+    Shape* toReturn;    // Shape pointer to be returned by the function
+
+    // Extract the pen
+
+    // If the shape being read in isn't a line, we know it needs a brush
+    if(shapeType != jconstants::LINE && shapeType != jconstants::POLYLINE)
+    {
+        // Extract brush
+
+        toReturn = constructShape(shapeType, dimensionsList);
+    }
+    else
+    {
+        toReturn = constructLine(shapeType, dimensionsList);
+    }
+
+    return toReturn;
+}
+
+Shape *ShapeData::extractText(QTextStream &qin, QList<int> &dimensionsList)
+{
+    QString body;   // Body of the text
+    QFont font; // Font of the text
+
+    // Get to the next whitspace, then read in the line as the text's body
+    getTo(qin, jconstants::TITLE_INTERRUPTOR);
+    body = qin.readLine();
+
+    // Extract the text's font
+    extractFont(qin, font);
+
+    return new Text(QPoint(dimensionsList[0], dimensionsList[1]), body, font);
+}
+
+void ShapeData::extractFont(QTextStream& qin, QFont& font)
+{
+    int tempInt; // Temporary storage for an integer
+    QString tempString; // Temporary storage for a string
+
+    // Get text size
+    getTo(qin, jconstants::TITLE_INTERRUPTOR);
+    qin >> tempInt;
+    font.setPointSize(tempInt);
+
+    // Get font family
+    getTo(qin, jconstants::TITLE_INTERRUPTOR);
+    tempString = qin.readLine();
+    font.setFamily(tempString);
+
+    // Get font style
+    getTo(qin, jconstants::TITLE_INTERRUPTOR);
+    qin >> tempString;
+    font.setStyle(stringToEnum(tempString, jconstants::TEXT_STYLE_STR, jconstants::TEXT_STYLE_VAL));
+
+    // Get font weight
+    getTo(qin, jconstants::TITLE_INTERRUPTOR);
+    qin >> tempString;
+    font.setWeight(stringToEnum(tempString, jconstants::TEXT_WEIGHT_STR, jconstants::TEXT_WEIGHT_VAL));
+}
+
+Shape* ShapeData::constructLine(QString& type, QList<int>& dimensions)
+{
+    if(type == jconstants::LINE)
+    {
+       return new Line(QPoint(dimensions[0], dimensions[1]), QPoint(dimensions[2], dimensions[3]));
+    }
+    else
+    {
+        return new Polyline(getPoints(dimensions), uint32_t(dimensions.size() / 2));
+    }
+}
+
+Shape* ShapeData::constructShape(QString& type, QList<int>& dimensions)
+{
+    if(type == jconstants::POLYGON)
+    {
+        return new Polygon(getPoints(dimensions), uint32_t(dimensions.size() / 2));
+    }
+    else if(type == jconstants::RECTANGLE)
+    {
+        return new Rect(QPoint(dimensions[0], dimensions[1]), dimensions[2], dimensions[3]);
+    }
+    else if(type == jconstants::SQUARE)
+    {
+        return new Rect(QPoint(dimensions[0], dimensions[1]), dimensions[2], dimensions[2]);
+    }
+    else if(type == jconstants::ELLIPSE)
+    {
+        return new Ellipse(QPoint(dimensions[0], dimensions[1]), dimensions[2], dimensions[3]);
+    }
+    else
+    {
+        return new Circle(QPoint(dimensions[0], dimensions[1]), dimensions[2]);
+    }
```

Comment:
That was also @ChrisRzech's suggestion.  I figured it'd be redundant since `extractShape()` already throws an exception `if(jconstants::TYPE_LIST.indexOf(shapeType) < 0)` and we never make it into these functions if that condition is true